### PR TITLE
[YW-299] feat(connector): Postgres/SQL data-source connector

### PR DESCRIPTION
**Pull request by youhaowei**
## Summary

- Adds `@dashframe/connector-postgres` — a read-only Postgres data-source connector mirroring `connector-rest` structure
- `PostgresConnector extends RemoteApiConnector`, `id = "postgres"`, auth-blind (no credential arg on `connect()`/`query()`)
- Two server-side mutations added to `app-artifacts.ts`: `listPostgresTables`, `queryPostgresTable`
- Registered in `ConnectorSetup.tsx` with a throwing renderer resolver (static metadata only)

Tracked internally as YW-299.

## Security architecture — needs-human review

The ticket spec described Layer 1 (`SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY`) as "the hard guard that rejects DML/DDL at the protocol level." **This is false.** That GUC sets the *default* transaction mode — an explicit `BEGIN READ WRITE` overrides it. This was empirically proven against Postgres 15.18 during the local review gate: `SELECT 1; COMMIT; BEGIN READ WRITE; CREATE TABLE pwned(...)` landed a write.

**What the PR actually ships:**

1. **Extended query protocol**: User SQL is sent as `client.query(sql, [])` — the empty params array forces Postgres to use Parse/Bind/Execute. Postgres rejects multi-statement SQL during Parse in this protocol, closing the `SELECT 1; BEGIN READ WRITE; …` bypass. This is a real code-level guard.
2. **`SET statement_timeout = '30s'`**: Bounds `pg_sleep()` and similar long-running DoS vectors.
3. **Deployment requirement documented in the connector docstring**: The DSN *must* use a Postgres role with only SELECT grants. `pg_read_file`, `dblink`, and superuser functions bypass all statement-level controls — there is no code guard for these.

The connector's read-only guarantee is strong when deployed with a least-privilege role. Without that operational requirement being met, the code-level guards are defense-in-depth, not a hard boundary. **This needs explicit owner acknowledgment before merge** — either update the spec to reflect the actual security model (operational floor + code guards), or scope down the ticket's promise.

## Test plan

- [x] 38 connector tests — all 5 ACs covered:
  - AC1: non-SELECT query rejected before any `client.query` call (0 spy calls on rejection)
  - AC2: `SET SESSION CHARACTERISTICS` is first call; `SET statement_timeout` is second; extended-protocol params contract (`[]`)
  - AC3: `connectionStringRef` is a SecretRef, never a DSN; fail-closed on empty resolver
  - AC4: `quoteIdentifier` escapes embedded `"` in schema/table; `$1` param for introspection values
  - AC5: `arrowBuffer`/`fieldIds`/`fields`/`rowCount` shape + Arrow deserialization
- [x] Full-graph typecheck: 48/48 tasks
- [x] Full-graph tests: 42/42 tasks (309 tests total, 271 server + 38 connector)
- [x] Local `code-review --effort high`: security BLOCK resolved (extended protocol + statement_timeout + deployment doc); correctness SHIP; api-contract SHIP

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01EeGZooqvrSHBxaTvTWkjJn

<details open><summary><h3>Greptile Summary</h3></summary>

This PR introduces `@dashframe/connector-postgres`, a read-only Postgres data-source connector that mirrors the existing `connector-rest` and `connector-notion` structure. Two server-side mutations (`listPostgresTables`, `queryPostgresTable`) are wired through a bound `SecretResolver` so no plaintext DSN ever reaches the call site.

- **Security architecture**: Two independent read-only guards — Layer 1 (`SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY`) and Layer 2 (keyword allowlist + extended query protocol via `queryMode: "extended"`, confirmed supported since pg 8.10.0 and used here at 8.22.0). An integration exploit test empirically validates that the multi-statement `BEGIN READ WRITE` bypass is rejected at Parse phase.
- **SQL sink protection**: Schema/table identifiers are always passed through `quoteIdentifier()`; user queries are sent with zero string interpolation; pagination limit/offset for the table-ref path is pushed down as `$1`/`$2` bind parameters, not concatenated.
- **Renderer safety**: `pg` is dynamically imported inside the auth callback, keeping Node-only code out of the browser bundle; the renderer-side connector instance uses a throwing resolver and the `showPostgres` feature flag remains `false` until the connect UI is complete.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

The new connector is safe to merge — the two read-only guards are independently load-bearing, all SQL sinks are parameterized or identifier-quoted, credentials never escape the auth callback, and the feature flag keeps the UI hidden until the connect flow is wired.

The two comments are minor quality observations: a redundant SET search_path round trip in listTablesInSchema and a usability edge case where nested block comments in user SQL produce a false rejection. Neither affects correctness or security. The exploit integration test empirically validates the key security property. All other paths — credential isolation, SQL sink guarding, Arrow serialization, and registry gating — are well-structured and consistent with existing codebase patterns.

No files require special attention. The only reviewer action needed is deciding whether to strip the redundant SET search_path TO call in listTablesInSchema before merge.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/connector-postgres/src/connector.ts | Core connector implementation — security guards (extended protocol, Layer 1/2), quoteIdentifier sink protection, and Arrow serialization all look correct. Minor: SET search_path in listTablesInSchema is redundant; nested block comments produce false rejections. |
| apps/server/src/functions/app-artifacts.ts | New Postgres mutations mirror Notion pattern correctly — auth-blind via bound resolver, defaultSchema validated as string before passing through, limit bounds checked before constructing pagination. |
| packages/connector-postgres/src/connector.test.ts | Comprehensive 38-test suite covering all 5 ACs; AC1 comment contradiction noted in previous thread was fixed in 705232bd. |
| packages/connector-postgres/src/__exploit.realpg.test.ts | Integration test that empirically validates extended-protocol multi-statement rejection and write side-effect absence; auto-skips without PG_DSN so CI remains green. |
| packages/app/src/components/providers/ConnectorSetup.tsx | Postgres connector registered with a throwing renderer resolver for static metadata only; dynamic pg import prevents Node-only code entering the browser bundle. |
| packages/app/src/lib/connectors/registry.ts | Feature-flag gating for postgres connector (showPostgres defaults false) mirrors existing Notion pattern correctly. |
| packages/connector-rest/src/connector.ts | Only change is updating the DANGEROUS_KEYS comment to include the empirically verified apache-arrow 21.1.0 behavior (silent column drop, not throw). |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client as Browser/Renderer
    participant Mutation as Server Mutation
    participant ConnectorFactory as postgresConnectorFor
    participant Vault as SecretVault
    participant Connector as PostgresConnector
    participant PG as Postgres DB

    Client->>Mutation: queryPostgresTable(dataSourceId, databaseId, tableId, limit)
    Mutation->>ConnectorFactory: postgresConnectorFor(ctx, dataSourceId)
    ConnectorFactory->>Vault: mintBoundResolver(vault, connectionString ref)
    ConnectorFactory-->>Mutation: connector (bound resolver, no DSN)

    alt databaseId is user SQL
        Mutation->>Connector: "connector.query(databaseId, tableId, {limit})"
        Connector->>Connector: assertReadOnlyQuery(sql) [Layer 2 allowlist]
        Connector->>Vault: this.auth(callback) resolves DSN
        Vault-->>Connector: plaintext DSN (callback scope only)
        Connector->>PG: SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY
        Connector->>PG: "SET statement_timeout = 30s"
        Connector->>PG: "query({text: sql, queryMode: extended}) [extended protocol]"
        PG-->>Connector: rows
    else databaseId is schema.table ref
        Connector->>PG: "SELECT * FROM schema.table LIMIT $1 OFFSET $2"
        PG-->>Connector: rows
    end

    Connector->>Connector: inferFieldsFromRows then Arrow IPC base64
    Connector-->>Mutation: arrowBuffer, fieldIds, fields, rowCount
    Mutation-->>Client: PostgresQueryResult (no DSN, no live DataFrame)
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client as Browser/Renderer
    participant Mutation as Server Mutation
    participant ConnectorFactory as postgresConnectorFor
    participant Vault as SecretVault
    participant Connector as PostgresConnector
    participant PG as Postgres DB

    Client->>Mutation: queryPostgresTable(dataSourceId, databaseId, tableId, limit)
    Mutation->>ConnectorFactory: postgresConnectorFor(ctx, dataSourceId)
    ConnectorFactory->>Vault: mintBoundResolver(vault, connectionString ref)
    ConnectorFactory-->>Mutation: connector (bound resolver, no DSN)

    alt databaseId is user SQL
        Mutation->>Connector: "connector.query(databaseId, tableId, {limit})"
        Connector->>Connector: assertReadOnlyQuery(sql) [Layer 2 allowlist]
        Connector->>Vault: this.auth(callback) resolves DSN
        Vault-->>Connector: plaintext DSN (callback scope only)
        Connector->>PG: SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY
        Connector->>PG: "SET statement_timeout = 30s"
        Connector->>PG: "query({text: sql, queryMode: extended}) [extended protocol]"
        PG-->>Connector: rows
    else databaseId is schema.table ref
        Connector->>PG: "SELECT * FROM schema.table LIMIT $1 OFFSET $2"
        PG-->>Connector: rows
    end

    Connector->>Connector: inferFieldsFromRows then Arrow IPC base64
    Connector-->>Mutation: arrowBuffer, fieldIds, fields, rowCount
    Mutation-->>Client: PostgresQueryResult (no DSN, no live DataFrame)
```

</a>
</details>

<sub>Reviews (3): Last reviewed commit: ["fix(connector-postgres): pagination push..."](https://github.com/youhaowei/dashframe/commit/e8aea190daf235aed6f02e08abdd2be7ab3d81b3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39739322)</sub>

### CODE REVIEW

File: packages/connector-postgres/src/connector.ts [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+ * Bind → Execute). Postgres rejects multi-statement SQL during Parse in this
+ * protocol, closing the `SELECT 1; BEGIN READ WRITE; …` bypass that would
+ * otherwise let an explicit transaction override the session-level read-only
+ * default set by SET SESSION CHARACTERISTICS.
+ */
+async function runUserQuery(
+  client: PgClientLike,
+  sql: string,
+): Promise<Record<string, unknown>[]> {
+  const result = await client.query(sql, []);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts]: Fixed in 705232bd. `runUserQuery` now uses `client.query({ text: sql, queryMode: "extended" })` — confirmed against `pg/lib/query.js` source: `requiresPreparation()` returns `true` for `queryMode === 'extended'`, closing the multi-statement bypass. The `PgClientLike` interface has a new `PgQueryConfig` overload. Tests updated to verify the config-object form reaches pg with `queryMode: "extended"`.

File: packages/connector-postgres/src/connector.ts [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    options?: QueryOptions,
+  ): Promise<ConnectorQueryResult> {
+    const schema = this.#config.defaultSchema ?? "public";
+
+    // Determine if databaseId is a "schema.table" reference (from connect())
+    // or a raw user SQL statement. This check happens BEFORE auth/client
+    // construction: a non-SELECT user query is rejected immediately (Layer 2
+    // fast-fail) with zero client.query calls — the error surfaces without
+    // touching the wire or the vault.
+    const isTableRef = /^[^(\s;]+\.[^(\s;]+$/.test(databaseId.trim());
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts]: Fixed in commit d4f1cf01 (earlier in this PR). The `isTableRef` regex was replaced with plain string `.includes()` checks (no backtracking): `!trimmed.includes(" ")`, `!trimmed.includes(";")`, `!trimmed.includes("(")`. CodeQL confirmed success on the new run.

File: packages/connector-postgres/src/connector.test.ts [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+  it("SELECT query reaches pg after SET SESSION", async () => {
+    await connector.query("SELECT count(*) FROM users", crypto.randomUUID());
+    const calls = spyClient.spy.mock.calls as Array<[string]>;
+    const queries = calls.map((c) => c[0].toUpperCase());
+    expect(queries[0]).toContain("SET SESSION CHARACTERISTICS");
+    expect(queries.some((q) => q.includes("SELECT"))).toBe(true);
+  });
+
+  it("WITH query passes the allowlist", async () => {
+    spyClient = makeSpyClient([{ n: 1 }]);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.test.ts]: Fixed in 705232bd. The contradictory comment was removed. The AC1 test now reads: 'The allowlist check runs BEFORE the auth/client path is entered — so even the SET SESSION statement is never issued. Zero pg calls total.'

File: packages/connector-postgres/src/connector.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts]: Acknowledged. This is a real round-trip bug: `connect()` lists `public.order items` (with a space), `query()` routes it through `isTableRef` which rejects it (has a space), then passes it to `assertReadOnlyQuery` which also rejects it (not a SELECT). The clean fix requires provenance-based routing from `connect()` — when a databaseId came from connect(), skip the isTableRef/SQL path and go directly to fetchTable. This is too large for this PR (touches the engine AnyConnector contract). Filing as a follow-up ticket.

File: apps/server/src/functions/app-artifacts.ts [hunk 4]
```diff
@@ -1193,6 +1194,119 @@ const queryNotionDatabase = mutation({
...
+  const auth = mintBoundResolver(
+    vault,
+    config.connectionString,
+    `DataSource(${dataSourceId})`,
+  );
+  // connectionStringRef is for introspection only; the bound `auth` resolver
+  // is the actual credential source. Pass the ref as-is — never coerce to "".
+  return makePostgresConnector(auth, {
+    connectionStringRef: config.connectionString,
+    defaultSchema: config.defaultSchema,
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/app-artifacts.ts]: Fixed in 705232bd. Added `typeof config.defaultSchema === 'string'` guard in `postgresConnectorFor` before passing to `makePostgresConnector` — non-string values from the JSONB blob are coerced to `undefined`, falling back to the 'public' default.

**Comment by youhaowei** [comment]
## Triage of open review findings

### Fixed in this PR (705232bd)
- **P1 (Codex) — multi-statement bypass**: `runUserQuery` now uses `client.query({ text: sql, queryMode: 'extended' })`. Confirmed against `pg/lib/query.js` source: `requiresPreparation()` returns `true` only for `queryMode === 'extended'`; an empty `[]` values array does NOT force the protocol (length check is `values.length > 0`).
- **P2 (Codex) — UI card no-op**: Postgres card is now behind a `showPostgres` feature flag (default `false`), mirroring the `showNotion` gate. Hidden until the renderer-side connect() flow is wired.
- **P2 (Greptile/Codex) — `defaultSchema` sink guard**: `postgresConnectorFor` now validates `typeof config.defaultSchema === 'string'` before passing to `makePostgresConnector`.
- **P2 (Greptile) — AC1 contradictory comment**: Fixed. Comment now accurately says zero pg calls are made.
- **CodeQL — polynomial regex**: The `isTableRef` regex was already replaced with plain string `.includes()` checks (earlier commit). CodeQL confirmed green.

### Deferred — follow-up tickets needed
- **Pagination pushdown** (Greptile P1, now outdated): `fetchTable` and `queryPostgresTable` materialize full result then slice. Correct fix: `LIMIT $1 OFFSET $2` pushed into the SQL, with the values passed via the extended protocol. Scope is larger than this PR (touches pagination API).
- **Quoted-table round-trip** (Codex P2): `connect()` returns IDs like `public.order items` (with space), but `query()`'s `isTableRef` check rejects IDs with spaces, sending them to `assertReadOnlyQuery` which also rejects. Fix needs provenance-based routing (touches engine AnyConnector contract).
- **NUMERIC/BIGINT coercion** (Codex P2): node-postgres returns `NUMERIC`/`BIGINT` as strings; `inferFieldsFromRows` marks them as `number` but Arrow receives the string. Needs explicit type coercion at the Arrow boundary.
- **Empty-table schema loss** (Codex P2): When a query returns zero rows, `fields` and `fieldIds` are empty `[]`. Needs pg `RowDescription` field metadata from query result even on empty result sets.
- **Arrow-reserved column names** (Codex P2): Column names like `__proto__`/`constructor` flow into `tableFromArrays` unfiltered.
- **`arrowBuffer` sensitivity gate** (Greptile P2): row data bypasses the sensitivity classifier. Same pattern as REST/Notion — this belongs at the sink (server enforcement layer), not the connector.

These are tracked as follow-ups and do not block the security invariants (ACs 1–5 are met).

File: packages/connector-postgres/src/connector.ts [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+ */
+async function fetchTable(
+  client: PgClientLike,
+  schema: string,
+  table: string,
+): Promise<Record<string, unknown>[]> {
+  const sql = `SELECT * FROM ${quoteIdentifier(schema)}.${quoteIdentifier(table)}`;
+  const result = await client.query(sql);
+  return result.rows;
+}
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:519]: Fixed in e8aea190. `fetchTable` now pushes the window into SQL: `SELECT * FROM "schema"."table" LIMIT $1 OFFSET $2` with limit/offset bound as value params (zero interpolation). When no limit is requested it emits the plain SELECT unchanged. Verified against live PG 15.18: a 5-row table queried with offset:1 limit:2 returns exactly 2 rows (the DB applied the window — no full materialization). Unit test asserts the emitted SQL carries `LIMIT $1 OFFSET $2` with params `[50, 20]`.

Note: the user-SQL path is deliberately NOT wrapped to inject LIMIT — wrapping (`SELECT * FROM (<user sql>) ...`) would re-interpolate user text (breaking the zero-interpolation Sink 3 contract) and break legal allowlisted forms (EXPLAIN/TABLE/trailing-semicolon). For user SQL the author bounds the query; statement_timeout + an in-process slice are the backstops.

File: packages/connector-postgres/src/connector.ts [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+        // User-supplied SQL — allowlist already passed above; send with zero interpolation.
+        // Sink 3: query sent with zero interpolation.
+        rows = await runUserQuery(client, databaseId);
+      }
+
+      // Apply pagination window if requested.
+      const limit = options?.pagination?.limit;
+      const offset = options?.pagination?.offset ?? 0;
+      const slicedRows =
+        limit !== undefined ? rows.slice(offset, offset + limit) : rows;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts]: Resolved by e8aea190. The call site (queryPostgresTable in app-artifacts.ts) already threaded `{ pagination: { offset, limit } }` into connector.query(); the gap was that fetchTable ignored it. Now fetchTable receives limit/offset and pushes `LIMIT $1 OFFSET $2` into the SQL. End-to-end pushdown confirmed against live PG.

File: packages/connector-postgres/src/connector.ts [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+        // Table reference — split on the first dot.
+        const dotIdx = databaseId.indexOf(".");
+        const refSchema = databaseId.slice(0, dotIdx);
+        const refTable = databaseId.slice(dotIdx + 1);
+        // Sink 2: both parts quoted via quoteIdentifier().
+        rows = await fetchTable(client, refSchema, refTable);
+      } else {
+        // User-supplied SQL — allowlist already passed above; send with zero interpolation.
+        // Sink 3: query sent with zero interpolation.
+        rows = await runUserQuery(client, databaseId);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts]: Fixed in e8aea190 — preview limits are now pushed into the SQL via `LIMIT $1 OFFSET $2` (table-ref path). See the detailed reply on the fetchTable thread.

File: packages/connector-postgres/src/connector.ts [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+      // Build Arrow column arrays from the inferred field set.
+      const columnArrays: Record<string, unknown[]> = Object.create(
+        null,
+      ) as Record<string, unknown[]>;
+      for (const field of fields) {
+        const colName = field.columnName ?? field.name;
+        columnArrays[colName] = slicedRows.map((r) => r[colName] ?? null);
+      }
+
+      const arrowTable = tableFromArrays(columnArrays);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:446]: Fixed in e8aea190. `inferFieldsFromRows` now drops `__proto__`/`constructor`/`prototype` column names before Arrow serialization (DANGEROUS_COLUMN_NAMES set), mirroring the REST connector's DANGEROUS_KEYS. Verified against live PG: a table with real `__proto__` and `constructor` columns yields fields `[id]` — no crash, no silent column loss. Note: I empirically checked apache-arrow 21.1.0 — these names do NOT crash tableFromArrays; `__proto__` silently DROPS the sibling real columns. Corrected the (inaccurate 'crash') comment in both this connector and connector-rest.

File: packages/connector-postgres/src/connector.ts [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+      const ipcBuffer = tableToIPC(arrowTable);
+      const base64 = Buffer.from(ipcBuffer).toString("base64");
+
+      return {
+        arrowBuffer: base64,
+        fieldIds: fields.map((f) => f.id),
+        fields,
+        rowCount: slicedRows.length,
+      };
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:455]: Investigated with the full code path traced. Summary: the *mechanism* in this finding is inaccurate, but it points at a real gap that is **product-wide, not Postgres-specific**, so the correct resolution is escalate-not-patch-here.

**Traced path (arrowBuffer from this connector):**
- `PostgresConnector.query()` returns `{ arrowBuffer, fieldIds, fields, rowCount }` — byte-for-byte the same `ConnectorQueryResult` shape as connector-rest and connector-notion. All three build fields via `createFieldsFromColumns` with NO sensitivity set (→ `unclassified`). Postgres does NOT diverge from the other connectors here.
- The server route `queryPostgresTable` (app-artifacts.ts) is a pass-through — returns arrowBuffer to the renderer, no gate. Same as `queryNotionDatabase`.
- The arrowBuffer does NOT reach the Parquet/DuckDB at-rest cache. `ParquetCache.write()` (engine-server/src/parquet-cache.ts) DOES enforce `makeSensitivityCacheWriteGate` (cleared-only) — but it gates **DuckDB query RESULTS over already-registered in-memory tables**, not connector output. The native engine's `registerArrowTable` is documented as in-memory-only ('row data is NEVER serialized to the filesystem').
- The one ungated at-rest write is `BrowserDataFrame.create → persistArrowData` → **IndexedDB** (engine-browser/src/dataframe.ts). That writes raw Arrow bytes with no sensitivity gate — and it is **uniform across ALL RemoteApiConnectors** (Notion hits it identically today).

**Postgres-specific fact:** there is no `materializePostgresTable` wired yet, so the Postgres arrowBuffer reaches NO at-rest sink in this PR at all (the connect UI is gated behind `showPostgres=false`).

**Disposition:** Adding a Postgres-only sensitivity gate would break the 'mirror REST/Notion' requirement and ship a per-connector fork of a product-wide concern. The honest fix is a connector-agnostic gate on the IndexedDB write path (`persistArrowData`). Filing that as a separate product-wide issue and surfacing to the owner. NOT a blocker for this connector PR — Postgres is not introducing or worsening the gap.

File: packages/app/src/components/providers/ConnectorSetup.tsx [hunk 10]
```diff
@@ -32,10 +33,23 @@ const notionConnectorForRegistry = makeNotionConnector(() => {
...
+    "[connector-registry] connect()/query() must not be called from the renderer — " +
+      "use the WyStack server mutations instead.",
+  );
+}, {});
+
 // Register connectors at module scope — synchronous, before any render.
 // getConnectorById() calls from any component on first render will resolve.
 registerConnector(localFileConnector);
 registerConnector(notionConnectorForRegistry);
+registerConnector(postgresConnectorForRegistry);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/components/providers/ConnectorSetup.tsx:52]: Resolved (commit 705232bd, earlier in this PR). `getConnectors()` now gates `connector.id === "postgres"` behind a `showPostgres` flag defaulting to `false`, mirroring the existing `showNotion` gate. Threaded through AddConnectionPanel + DataPickerContent. The Postgres card is hidden until the renderer-side connect() flow is wired. Registry tests pass (32/32).

File: packages/connector-postgres/src/connector.ts [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+      // Infer fields from rows.
+      const fields = inferFieldsFromRows(slicedRows, tableId);
+
+      // Build Arrow column arrays from the inferred field set.
+      const columnArrays: Record<string, unknown[]> = Object.create(
+        null,
+      ) as Record<string, unknown[]>;
+      for (const field of fields) {
+        const colName = field.columnName ?? field.name;
+        columnArrays[colName] = slicedRows.map((r) => r[colName] ?? null);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:443]: Acknowledged, filing as a follow-up. node-postgres returns NUMERIC/BIGINT as strings; `inferFieldsFromRows` may mark them `number` while Arrow receives the string, causing a type mismatch. The clean fix is explicit value coercion at the Arrow boundary (the same shape as REST's coerce step). Out of scope for this PR's security/memory focus — tracked as a follow-up ticket.

File: packages/connector-postgres/src/connector.ts [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+      }
+
+      // Apply pagination window if requested.
+      const limit = options?.pagination?.limit;
+      const offset = options?.pagination?.offset ?? 0;
+      const slicedRows =
+        limit !== undefined ? rows.slice(offset, offset + limit) : rows;
+
+      // Infer fields from rows.
+      const fields = inferFieldsFromRows(slicedRows, tableId);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:435]: Acknowledged, filing as a follow-up. When a query returns zero rows, fields are inferred from rows so `fields`/`fieldIds` come back empty even when the table has a real schema. The fix needs pg's RowDescription column metadata from the query result (available even on empty result sets) rather than row-based inference. Out of scope here — tracked as a follow-up ticket.

File: packages/connector-postgres/src/connector.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:443]: Correction on ownership: this is tracked as a follow-up by the coordinator (not filed by me). Out of scope for this PR's security/memory-safety focus. Resolving the thread.

File: packages/connector-postgres/src/connector.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-postgres/src/connector.ts:435]: Correction on ownership: this is tracked as a follow-up by the coordinator (not filed by me). Out of scope here. Resolving the thread.

**Comment by youhaowei** [comment]
## Final status — ready for human review (do not auto-merge)

**Head:** `e8aea190` · **CI:** all green on this head (Lint/CodeQL/Analyze/E2E/Greptile/CodeRabbit) · **Merge state:** CLEAN · **Open non-outdated review threads:** 0

### The merge-blocking P1 (multi-statement read-only bypass) is closed and PROVEN against live Postgres
The earlier `client.query(sql, [])` did NOT force the extended protocol (pg checks `values.length > 0`; `[]` stays on the simple protocol). Fixed to `client.query({ text, queryMode: "extended" })`. A real-PG regression test (`__exploit.realpg.test.ts`, auto-skips without `PG_DSN`) drives the actual connector against live PG 15.18. Discriminating evidence, same SET SESSION read-only default on both:
| | result |
|---|---|
| OLD `client.query(sql, [])` | bypass **ACCEPTED**, `pwned` table **CREATED** |
| NEW `queryMode: "extended"` | rejected (error 42601 "cannot insert multiple commands"), no table |

### Also fixed this round
- **Pagination pushdown** (memory safety): table-ref path now pushes `LIMIT $1 OFFSET $2` (bound params) into SQL instead of materializing the full table then slicing. Verified end-to-end against live PG.
- **Arrow-reserved column filter**: `__proto__`/`constructor`/`prototype` dropped before serialization (mirrors REST). Verified against live PG (real columns with those names → no crash, no silent column loss).
- **defaultSchema sink guard** + **Postgres card feature-gated** (`showPostgres` default-false) + AC1 comment fix.

### Needs-human before merge (not blockers introduced by this PR)
1. **Connect UI is unwired** — the Postgres card is intentionally hidden behind `showPostgres=false`; there is no `materializePostgresTable` renderer flow yet. The server data-plane routes work; the renderer add-connection flow is a follow-up.
2. **Security-spec reframe** — Layer 1 (`SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY`) sets the session DEFAULT; it is not a protocol-level guard. The actual guarantee against the multi-statement bypass is `queryMode: "extended"` (Parse-time rejection). Documented in the connector docstring; worth a maintainer eye on the wording.
3. **Product-wide privacy gap (escalated)** — raw row data reaches at-rest IndexedDB via `persistArrowData` with no `Field.sensitivity` gate. This is **uniform across all RemoteApiConnectors** (REST/Notion identical), NOT Postgres-specific, and Postgres reaches no at-rest sink at all in this PR. The correct fix is a connector-agnostic gate at the IndexedDB sink — the coordinator is filing the follow-up. A Postgres-only gate was deliberately NOT added (would fork a product-wide concern).

Follow-ups owned by the coordinator: NUMERIC/BIGINT coercion, empty-result schema preservation, quoted-table-name round-trip, the IndexedDB sensitivity gate.
